WooCommerce Analytics: Queue `add_to_cart` Events

greenafrican opened the pull request:
Fixes #9049 

#### Changes proposed in this Pull Request:

* move to queueing all `add_to_cart` events and logging on the next page
* prevent duplicate `product_view` events from product page

#### Testing instructions:

* click any add to cart button and look for event to be logged in the next page load
* add a link to a post or page with href like `http://exampleshop.com/?add-to-cart=${product_id}`
* make sure page view and cart event logged in next page load

<!-- Add the following only if this is meant to be in changelog -->
#### Proposed changelog entry for your changes:

greenafrican commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -325,4 +320,59 @@ public function get_user_id() {
+		 * @param $cart_item_data
+		 */
+		public function capture_add_to_cart( $cart_item_key, $product_id, $quantity, $variation_id, $variation, $cart_item_data ) {
+			if ( ! is_single() ) {
```
`is_single()` doesn't seem to work here so we always get a `page_view` even if we've already logged one on the single product page. Any ideas @psealock @timmyc ?

timmyc commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -73,61 +72,37 @@ public function wp_head_bottom() {
-		);
+		// check for previous add-to-cart cart events
+		$data = WC()->session->get( 'wca_session_data' );
+		if ( !empty( $data ) ) {
```
minor but space between `! empty`

timmyc commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -325,4 +303,89 @@ public function get_user_id() {
 	}
 
+	/**
+		 * @param $cart_item_key
```
Indentation is a little wonky right here.

timmyc commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -325,4 +303,89 @@ public function get_user_id() {
+		public function capture_add_to_cart( $cart_item_key, $product_id, $quantity, $variation_id, $variation, $cart_item_data ) {
+			$referer_postid = isset( $_SERVER['HTTP_REFERER'] ) ? url_to_postid( $_SERVER['HTTP_REFERER'] ) : 0;
+			// if the referring post is not a product OR the product being added is not the same as post
+			// (eg. related product list on single product page) then include a product view event
```
Also noting from our convo today - how will this bit of logic respond to a quick add event from a product listing page?

justinshreve replied:
Good point. It captures the `add_to_cart` action properly -- but it does also fire off a view which might not be accurate in this case.

psealock replied:
yes, adding a product from the product page results in 2 `product_view`s and one `add-to-cart`

>if the referring post is not a product OR the product being added is not the same as post
(eg. related product list on single product page) then include a product view event

This is correct, and when I step through the code, this portion behaves as expected. Digging further, the extra page view is from the page refresh, and you can see that working in the master branch.

@greenafrican Since we are utilizing `session`, in `capture_product_view` we can check to see if a `product_view` is about to be dropped. This may depend on timing. Is it worth exploring?

greenafrican replied:
> Digging further, the extra page view is from the page refresh

Yes, I had an extra check to prevent these but removed it yesterday as it can cause other issues down the line. For example, if the user continues to `add_to_cart` we log all the additional `add_to_cart` events but no further `product_views`. It is not ideal/perfect as is but I'd rather there were a few extra `product_views` than have a situation where there can be more `add_to_carts` than `product_views`. The more I think about this the more I wonder if we don't need to move to reporting on uniques rather than simple counts.

greenafrican replied:
> how will this bit of logic respond to a quick add event from a product listing page?

@timmyc It will add both a `product_view` and an `add_to_cart` event, which is what we want. 

> It captures the add_to_cart action properly -- but it does also fire off a view which might not be accurate in this case.

@justinshreve My thoughts are that we do want the `product_view` here for 2 reasons: 1) we don't want a situation where we have more `add_to_cart` events than `product_views` in a funnel, 2) I feel as though, by clicking `add_to_cart`, the user has 'looked' at the product.

timmyc commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -325,4 +303,89 @@ public function get_user_id() {
+
+	/**
+	 * @param $product_id
+	 * @param $qty
```
s/qty/quantity ?

timmyc commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -325,4 +303,89 @@ public function get_user_id() {
+
+		// extract new event data
+		$new_data = array(
+			'event'     		=> $event,
```
spacing a bit off here too.

timmyc commented:
/cc @justinshreve - would be keen to have you 👀 this PR if you have some time.

psealock commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -325,4 +303,89 @@ public function get_user_id() {
+		$product_id,
+		$quantity,
+		$event,
+		$cart_item_key = '' ) {
```
`$cart_item_key` is unused in this function

timmyc commented:
Any ideas why 88e8644 broke the build for this branch?

zinigor commented:
It probably just needs a restart - I restarted two failed jobs, let's see what happens.

timmyc commented:
Thanks @zinigor

greenafrican commented:
Thanks for the reviews @timmyc @psealock @justinshreve 

Fixes included in the latest squashed commit:
- spaces between `! empty` and `! array`
- indentation and spacing fixes
- rename `qty` to `quantity`
- moved arguments onto 1 line
- documented `$product` as an object, not array
- removed unused `$cart_item_key`

greenafrican commented:
@jeherve Would you mind reviewing this PR, to be included in v5.9.1 if possible?

greenafrican commented:
@psealock @timmyc @justinshreve would you mind doing another review of this?
